rig/machine_control/regions.py

```diff
@@ -0,0 +1,200 @@
+        When 3 is used the specified region will ONLY select the given chip,
+        for other regions surrounding chips will also be selected.
+
+    Returns
+    -------
+    int
+        A 32-bit value representing the co-ordinates of the chunk of SpiNNaker
+        chips that should be selected and the blocks within this chunk that are
+        selected.  As long as bits (31:16) are the same these values may be
+        OR-ed together to increase the number of sub-blocks selected.
+    """
+    shift = 6 - 2*level
+
+    bit = ((x >> shift) & 3) + 4*((y >> shift) & 3)  # bit in bits 15:0 to set
+
+    mask = 0xffff ^ ((4 << shift) - 1)  # in {0xfffc, 0xfff0, 0xffc0, 0xff00}
+    nx = x & mask  # The mask guarantees that bits 1:0 will be cleared
+    ny = y & mask  # The mask guarantees that bits 1:0 will be cleared
+
+    #        sig bits x | sig bits y |  2-bit level  | region select bits
+    region = (nx << 24) | (ny << 16) | (level << 16) | (1 << bit)
+    return region
+
+
+def reduce_regions_hierarchically(chips):
+    """Create a reduced set of regions by minimising a hierarchy tree.
+
+    Parameters
+    ----------
+    chips : {(x, y) ...}
+        Set of x and y co-ordinate pairs.
+
+    Returns
+    -------
+    generator
+        A generator which yields 32-bit region codes which minimally cover the
+        set of given chips.
+    """
+    t = RegionTree()
+    for (x, y) in chips:
+        t.add_coordinate(x, y)
+    return t.get_regions()
+
+
+def get_minimal_flood_fills(targets):
+    """Generate a reduced set of flood fill parameters.
+
+    Parameters
+    ----------
+    targets : {(x, y) : set([c, ...]), ...}
+        For each used chip a set of core numbers onto which an application
+        should be loaded.  E.g., the output of
+        :py:func:`~rig.place_and_route.util.build_application_map` when indexed
+        by an application.
+
+    Returns
+    -------
+    generator
+        A generator which yields region and core mask pairs indicating
+        parameters to use to flood-fill an application.  `region` and
+        `core_mask` are both integer representations of bit fields that are
+        understood by SCAMP.
+    """
+    # Build a dictionary mapping core mask -> chips where this should be
+    # applied.
+    cores_to_targets = collections.defaultdict(set)
+    for (x, y), cores in iteritems(targets):
+        # Build the core mask
+        core_mask = 0x0000
+        for c in cores:
+            core_mask |= 1 << c
+
+        # Add to the targets dict
+        cores_to_targets[core_mask].add((x, y))
+
+    # For each of these cores build the minimal set of regions
+    for core_mask, coordinates in iteritems(cores_to_targets):
+        regions = reduce_regions_hierarchically(coordinates)
+        for r in regions:
+            yield (r, core_mask)
+
+
+class RegionTree(object):
+    """Tree structure for use in minimising regions."""
```

Comment:
"""A tree structure for use in minimising sets of regions.
    
    A tree is defined which reflects the definition of SpiNNaker regions like so:
    The tree's root node represents a 256x256 grid of SpiNNaker chips. This grid is
    broken up into 64x64 grids which are represented by the (level 1) child nodes of
    the root.  Each of these level 1 nodes' 64x64 grids are broken up into 16x16
    grids which are represented by their (level 2) children. These level 2 nodes
    have their 16x16 grids broken up into 4x4 grids represented by their (level 3)
    children. Level 3 children explicitly list which of their sixteen cores are part
    of the region.
    
    If any of a level 2 node's level 3 children have all of their cores selected,
    these level 3 nodes can be removed and replaced by a level 2 region with the
    corresponding 4x4 grid selected. If multiple children can be replaced with level
    2 regions, these can be combined into a single level 2 region with the
    corresponding 4x4 grids selected, resulting in a reduction in the number of
    regions required. The same process can be repeated at each level of the
    hierarchy eventually producing a minimal set of regions.
    
    This data structure is specified by supplying a sequence of (x, y) coordinates
    of chips to be represented by a series of regions using
    :py:meth:`.add_coordinate`. This method minimises the tree during insertion
    meaning a minimal set of regions can be extracted by :py:meth:`.get_regions`
    which simply traverses the tree.
    """